Description: Find match stop at end or beginning of buffer
Author: Jens Peter Secher <jps@debian.org>

Index: sml-mode/sml-move.el
===================================================================
--- sml-mode.orig/sml-move.el	2010-09-11 15:27:15.000000000 +0200
+++ sml-mode/sml-move.el	2010-09-11 15:33:35.000000000 +0200
@@ -135,6 +135,7 @@
 	(condition-case () (forward-sexp 1) (error (forward-char 1))))
       (setq level
 	    (cond
+	     ((and (eobp) (> level 1)) (error "Unbalanced"))
 	     ((sml-looking-back-at this) (1+ level))
 	     ((sml-looking-back-at match) (1- level))
 	     (t (error "Unbalanced")))))
@@ -150,6 +151,7 @@
 	(condition-case () (backward-sexp 1) (error (backward-char 1))))
       (setq level
 	    (cond
+	     ((and (bobp) (> level 1)) (error "Unbalanced"))
 	     ((looking-at this) (1+ level))
 	     ((looking-at match) (1- level))
 	     (t (error "Unbalanced")))))
